nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/RunMongoAggregation.java

```diff
@@ -0,0 +1,203 @@
+            .displayName("Query")
+            .expressionLanguageSupported(true)
+            .description("The aggregation query to be executed.")
+            .required(true)
+            .addValidator(AGG_VALIDATOR)
+            .build();
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(CHARSET);
+        _propertyDescriptors.add(QUERY);
+        _propertyDescriptors.add(QUERY_ATTRIBUTE);
+        _propertyDescriptors.add(BATCH_SIZE);
+        _propertyDescriptors.add(RESULTS_PER_FLOWFILE);
+        _propertyDescriptors.add(SSL_CONTEXT_SERVICE);
+        _propertyDescriptors.add(CLIENT_AUTH);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_RESULTS);
+        _relationships.add(REL_ORIGINAL);
+        _relationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    static String buildBatch(List batch) {
+        ObjectMapper mapper = new ObjectMapper();
+        String retVal;
+        try {
+            retVal = mapper.writeValueAsString(batch.size() > 1 ? batch : batch.get(0));
+        } catch (Exception e) {
+            retVal = null;
+        }
+
+        return retVal;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+
+        String query = context.getProperty(QUERY).evaluateAttributeExpressions(flowFile).getValue();
+        String queryAttr = context.getProperty(QUERY_ATTRIBUTE).evaluateAttributeExpressions(flowFile).getValue();
+        Integer batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        Integer resultsPerFlowfile = context.getProperty(RESULTS_PER_FLOWFILE).asInteger();
+
+        Map attrs = new HashMap();
+        if (queryAttr != null && queryAttr.trim().length() > 0) {
+            attrs.put(queryAttr, query);
+        }
+
+        MongoCollection collection = getCollection(context);
+        MongoCursor iter = null;
+
+        try {
+            List<Bson> aggQuery = buildAggregationQuery(query);
+            AggregateIterable it = collection.aggregate(aggQuery);
+            it.batchSize(batchSize != null ? batchSize : 1);
+
+            iter = it.iterator();
+            List batch = new ArrayList();
+
+            while (iter.hasNext()) {
+                batch.add(iter.next());
+                if (batch.size() == resultsPerFlowfile) {
+                    writeBatch(buildBatch(batch), null, context, session, attrs, REL_RESULTS);
```

Comment:
Why hardcode the parent flowfile to null here? You can use a reference to flowFile and if it happens to be null, then it will work as it does here.

Reply:
That might have been a mistake on my part that came from doing the rebase against a recent change to GetMongo.